### ISSUE TITLE
Inferred Unknown in descriptor

### ISSUE THREAD

**Issue by mehdigmira** [microsoft/pyright#5778]
Pyright version 1.1.323

```python 

from __future__ import annotations

from typing import *

_T = TypeVar("_T")


class ORMDescriptor(Generic[_T]):
    @overload
    def __get__(self, instance: Any, owner: Literal[None]) -> ORMDescriptor[_T]:
        ...

    @overload
    def __get__(self, instance: Literal[None], owner: Any) -> Any:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T:
        ...

    def __get__(self, instance: object, owner: Any) -> Union[ORMDescriptor[_T], Any, _T]:
        ...


class A:
    y: ORMDescriptor[str]

    @property
    def label_descriptions(self):
        reveal_type(
            self.y
        )  # information: Type of "self.y" is "Unknown", but if i remove the first overload of __get__ in ORMDescriptor, it is "str"
```

pyright seems to wrongfully assume that self.y is Unknown. This passes under 1.1.320.

**Comment by erictraut** [comment]
Here's what's happening. When a descriptor is accessed at runtime, the value received by the `owner` parameter in the `__get__` method is either a class or `None` depending on whether the descriptor was accessed through the class or an instance of the class. Mypy doesn't model this correctly when evaluating an overloaded `__get__` method. In particular, it assumes that the value passed to the `owner` parameter is an `Any` rather than a `None`. In your code sample above, this makes the evaluation of the overload ambiguous, so mypy consequently evaluates the result as `Any`.

In pyright, I previously attempted to model this correctly so it was consistent with the runtime, but I found that a number of existing type stubs were sloppy in how they defined the `__get__` overload. This resulted in errors in pyright that didn't appear in mypy. I therefore changed pyright to match mypy's "sloppier" behavior. This works for most cases and masks the problems with incorrect `__get__` overloads, but in the code above it results in ambiguous overload resolution, and (like mypy), pyright produces an `Any`. But this is an implicit `Any`, which pyright refers to as `Unknown`. So that explains why you're seeing an `Unknown` appear in this case.

My inclination is to abandon my attempt to match mypy's modeling of descriptors and model it correctly as per the runtime behavior. This would force stub/library authors to fix their type information so the `__get__` overloads are defined correctly. 

Unfortunately, `mypy_primer` (a tool that measures the impact of a type checker change) indicates that this could be pretty painful for library authors. See [these results](https://github.com/microsoft/pyright/pull/5783). All of the green lines are new errors that result from this change.

@Gobot1234, it looks like this change would affect the `steam.py` library, which you maintain, so I'd like to get your input before I make this change. In your code, the `Chat._type_args` method is decorated with `@utils.classproperty`, but `classproperty` defines `__get__` in a way that means it won't work if invoked from an instance of class.

**Comment by Gobot1234** [comment]
I'd be happy to change the definition of `classproperty.__get__` if it's more consistent with runtime, I wasn't aware of this being incorrect (maybe cause everything does it the same way to match mypys behaviour if I had to guess). Do you have an example of how it should be annotated to not fail after the PR? I also don't know where the typing of descriptors is documented so it might be worth asking for such a feature to be documented so that there's at least some reference for how to do this.

**Comment by erictraut** [comment]
If the descriptor is meant to work when accessed through a class instance, then the `owner` parameter should accept a `None`. You could either use a single signature and use a union for the `owner` parameter (e.g. `owner: type | None`) or you could define an overload where one signature accepts a `None` and the other accepts a `type`.

**Comment by Gobot1234** [comment]
Is `type | None` not a subtype of Any?

**Comment by erictraut** [comment]
> Is type | None not a subtype of Any?

Yes, but `None` is not a subtype of `type[Any]`. In your current definition of `classproperty.__get__`, the second parameter (which you call `type` rather than `owner`) is annotated with `TT_co` which is a type variable with an upper bound of `type[Any]`.

**Comment by Gobot1234** [comment]
Ok so I've been looking into this with Rapptz (as he is also affected by this change) a bit because we both weren't aware of descriptors being callable in a 1-arg form. I've changed my definition of `classproperty` to
```py
class classproperty(Generic[TT_co, T_co]):
    def __init__(self, func: Callable[[TT_co], T_co]):
        self.__func__ = func

    @overload
    def __get__(self, instance: Any, owner: None = None) -> T_co: ...
    @overload
    def __get__(self, instance: None, owner: TT_co) -> T_co: ...
    def __get__(self, instance: Any, owner: TT_co | None = None) -> T_co:
        return self.__func__(type(instance) if owner is None else owner)
```
Unfortunately I don't see a good way to type this (without a silly little idea of an `InstanceOf` special form that I briefly toyed with for PEP 673 which would be useful as the return type of `type.__call__` or so other operations made sense in metaclasses)
```py
class classproperty(Generic[TT_co, T_co]):
    def __init__(self, func: Callable[[TT_co], T_co]):
        self.__func__ = func

    @overload
    def __get__(self, instance: InstanceOf[TT_co], owner: None = None) -> T_co: ...
    @overload
    def __get__(self, instance: None, owner: TT_co) -> T_co: ...
    def __get__(self, instance: Any, owner: TT_co | None = None) -> T_co:
        return self.__func__(type(instance) if owner is None else owner)
```

**Comment by erictraut** [comment]
Digging into this further, it looks like my understanding about the second parameter was incorrect. I thought that the runtime passed a value of `None` for the second argument if `__get__` is invoked through an instance, but it appears that it always passes the class even when `instance` is provided.

I'm not sure when (if ever) the runtime passes `None` as the second argument. Looking at the CPython code, it doesn't appear to do so in the normal attribute access code path. Maybe there's some other unusual case where it passes `None` as the second argument (or doesn't pass a second argument, whereby the default value of `None` would be used). Perhaps this is legacy behavior left over from Python 2.

Here's some test code that demonstrates what the Python 3 runtime is doing:

```python
class MyDescriptor:
    def __get__(self, instance, owner):
        print("MyDescriptor.__get__", instance, owner)
        return None

class MyMeta(type):
    y = MyDescriptor()

class A(metaclass=MyMeta):
    x = MyDescriptor()

A.x  # None, class A
A().x  # object A, class A
A.y  # class A, class MyMeta
```

I'll update pyright's code accordingly and see what that does in terms of compatibility with existing code bases.

@Gobot1234, I think you're struggling to annotate this because you're using a type variable with an upper bound of `type[Any]`. That's pretty unusual. Normally, you would define the TypeVar without an upper bound and use `type[T]` to indicate "a subclass of `type` that creates a `T` when its constructor is called".

```python
T_co = TypeVar("T_co")

class classproperty(Generic[T_co]):
    def __init__(self, func: Callable[[type[T_co]], T_co]):
        self.__func__ = func

    def __get__(self, instance: T_co | None, owner: type[T_co]) -> T_co:
        ...
```

**Comment by Gobot1234** [comment]
> Digging into this further, it looks like my understanding about the second parameter was incorrect. I thought that the runtime passed a value of `None` for the second argument if `__get__` is invoked through an instance, but it appears that it always passes the class even when `instance` is provided.

I was under the impression the interpreter handled it here
```py
A.__dict__["x"].__get__(A())
```
however, this doesn't work. It just seems to be that from the spec it says you can make the last argument not required/optional https://docs.python.org/3/reference/datamodel.html#object.__get__:~:text=PEP%20252%20specifies%20that%20__get__()%20is%20callable%20with%20one%20or%20two%20arguments.

> @Gobot1234, I think you're struggling to annotate this because you're using a type variable with an upper bound of `type[Any]`. That's pretty unusual. Normally, you would define the TypeVar without an upper bound and use `type[T]` to indicate "a subclass of `type` that creates a `T` when its constructor is called".

That would imply a relation between the type of the first parameter and the return type, however there is none, the code that uses this example looks like
```py
class Commentable:
    @classproperty
    def _TYPE(cls: type[Self]) -> CommentThreadType:
        return CommentThreadType[cls.__name__]
```
anyway I think this is besides the point, it doesn't seem like I should need to change the definition so my code should work fine.

**Comment by erictraut** [comment]
Thanks for that documentation reference. That provides the clarity I was looking for.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
